src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1962,6 +1970,8 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
             | Rvalue::UnaryOp(..)
             | Rvalue::Discriminant(..) => {}
         }
+
+        rvalue.ty(mir, tcx)
```

Comment:
...this is the case that bothers me. Now we have to know that *sometimes* we can use this, sometimes we can't...